File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"
#include <stdexcept>

namespace Bankaccount {
void Bankaccount::open() {
  if (saving != nullptr)
    throw std::runtime_error("");
  saving = new int(0);
}
Bankaccount::Bankaccount() {}
Bankaccount::~Bankaccount() { close(); }
int Bankaccount::balance() {
  if (saving == nullptr) {
    throw std::runtime_error("");
  }
  return *saving;
}
void Bankaccount::deposit(int d) {
  if (saving == nullptr || d <= 0)
    throw std::runtime_error("");
  lock.lock();
  *saving += d;
  lock.unlock();
}
void Bankaccount::withdraw(int w) {
  if (saving == nullptr || w <= 0 || *saving < w)
    throw std::runtime_error("");
  lock.lock();
  *saving -= w;
  lock.unlock();
}
void Bankaccount::close() {
  if (saving == nullptr)
    throw std::runtime_error("");
  delete saving;
  saving = nullptr;
}

} // namespace Bankaccount

File: solutions/go/bank-account/1/bank_account.go
package account

import "sync"

// Define the Account type here.
const (
	open  = true
	close = false
)

type Account struct {
	status  bool
	balance int64
	w       sync.Mutex
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	return &Account{status: open, balance: amount}
}

func (a *Account) Balance() (int64, bool) {
	if !a.status {
		return 0, false
	}
	return a.balance, open
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.w.Lock()
	if !a.status {
		a.w.Unlock()
		return 0, false
	}

	if a.balance+amount <= 0 {
		a.w.Unlock()
		return 0, false
	}

	a.balance += amount
	a.w.Unlock()
	return a.balance, open
}

func (a *Account) Close() (int64, bool) {
	if a.status {
		a.status = close
		return a.balance, true
	} else {
		return 0, false
	}
}
